src/vcpkg/commands.build.cpp

```diff
@@ -1488,55 +1488,84 @@ namespace vcpkg
-            static constexpr auto END_BLOCK_MAX_LENGTH = 45'000;
-            auto log = fs.read_contents(path, VCPKG_LINE_INFO);
-            if (log.size() > MAX_LOG_LENGTH)
+            auto first_block_end = log.find_last_of('\n', start_block_max_length);
+            if (first_block_end == std::string::npos)
             {
-                auto first_block_end = log.find_first_of('\n', START_BLOCK_LENGTH);
-                if (first_block_end == std::string::npos || first_block_end > START_BLOCK_MAX_LENGTH)
-                {
-                    first_block_end = START_BLOCK_LENGTH;
-                }
-
-                auto last_block_end = log.find_last_of('\n', log.size() - END_BLOCK_LENGTH);
-                if (last_block_end == std::string::npos || last_block_end < log.size() - END_BLOCK_MAX_LENGTH)
-                {
-                    last_block_end = log.size() - END_BLOCK_LENGTH;
-                }
-
-                auto first = log.begin() + first_block_end;
-                auto last = log.begin() + last_block_end;
-                auto skipped_lines = std::count(first, last, '\n');
-                log.replace(first, last, fmt::format("\n...\nSkipped {} lines\n...", skipped_lines));
+                first_block_end = start_block_max_length;
             }
 
-            while (!log.empty() && log.back() == '\n')
+            auto last_block_start = log.find_first_of('\n', log.size() - end_block_max_length);
+            if (last_block_start == std::string::npos)
             {
-                log.pop_back();
+                last_block_start = log.size() - end_block_max_length;
             }
 
-            return fmt::format("<details><summary>{}</summary>\n\n```\n{}\n```\n</details>", path.native(), log);
+            auto first = log.begin() + first_block_end;
+            auto last = log.begin() + last_block_start;
+            auto skipped_lines = std::count(first, last, '\n');
+            out.append(log.begin(), first);
+            fmt::format_to(std::back_inserter(out), skipped_msg.c_str(), skipped_lines);
+            out.append(last, log.end());
         }
-    };
+        else
+        {
+            out += log;
+        }
+
+        while (!out.empty() && out.back() == '\n')
+        {
+            out.pop_back();
+        }
+        Strings::append(out, details_end);
+    }
+
+    void append_logs(View<std::pair<Path, std::string>> logs, int max_size, std::string& out)
+    {
+        int extra_size = 0;
+        max_size -= static_cast<int>(logs.size()) * 2;
+        auto size_per_log = max_size / static_cast<int>(logs.size());
+        for (auto& entry : logs)
+        {
+            const auto old_size = out.size();
+            append_log(entry.first, entry.second, size_per_log + extra_size, out);
+            extra_size += size_per_log - static_cast<int>(out.size() - old_size);
+            if (out.size() != old_size)
+            {
+                out += "\n\n";
+            }
+        }
```

Comment:
Instead of trying to track how much "extra space" we have left (which I suspect is a lot of the reason to use int?) I think it would be clearer just to know the maximum value we should be at which is not dependent on the input:

```suggestion
        auto size_per_log = static_cast<size_t>(max_size) / logs.size();
        size_t maximum = 0;
        for (auto& entry : logs)
        {
            maximum += size_per_log;
            const auto old_size = out.size();
            const auto available = maximum - old_size;
            // append_log knows this bool so it would be nice if it returned it rather
            // than us comparing with old_size...
            append_log(entry.first, entry.second, available, out);
            if (out.size() != old_size)
            {
                out += "\n\n";
            }
        }
```

If you want you could also make `maximum = (idx + 1) * size_per_log`.

Reply:
I note that this will penalize long logs less if they are at the "end"; maybe it would be a good idea to sort them.